File: data_structure/LinkedList/java/singly_linked_list/LinkedList.java
package data_structure.linked_list.singly_linked_list;


public class LinkedList <T> {
    Node<T> head;
    private int count;

    public LinkedList() {
        this.head = null;
        this.count = 0;
    }

    /**
     * 원하는 위치에 요소 추가
     */
    public void insertAt(int index, T data) {

        if (index > count || index < 0) {
            throw new IndexOutOfBoundsException();
        }

        Node<T> newNode = new Node<>(data, null);

        if (index == 0) {
            newNode.setNext(head);
            head = newNode;
        } else {
            Node<T> currentNode = head;

            for (int i = 0; i < index - 1; i++) {
                currentNode = currentNode.getNext();
            }
            newNode.setNext(currentNode);
            currentNode.setNext(newNode);
        }

        count++;
    }

    /**
     * 원하는 위치의 요소 조회
     */
    public Node<T> getNodeAt(int index) {
        if (index >= count || index < 0)
            throw new IndexOutOfBoundsException();

        Node<T> currentNode = head;
        for (int i = 0; i < index; i++) {
            currentNode = currentNode.getNext();
        }

        return currentNode;
    }

    /**
     * 원하는 위치의 요소 삭제
     */
    public Node<T> deleteAt(int index) {
        if (index >= count || index < 0)
            throw new IndexOutOfBoundsException();

        Node<T> currentNode = head;

        if (index == 0) {
            Node<T> deleteNode = head;
            head = head.getNext();
            count--;

            return deleteNode;
        } else {
            for (int i = 0; i < index - 1; i++) {
                currentNode = currentNode.getNext();
            }

            Node<T> deleteNode = currentNode.getNext();
            currentNode = currentNode.getNext().getNext();
            count--;

            return deleteNode;
        }
    }

    /**
     * 전체 요소 출력
     */
    public void printAll() {
        Node<T> currentNode = head;
        String text = "[";

        for (int i = 0; i < count; i++) {
            text += String.valueOf(currentNode.getData());

            if (i != count - 1)
                text += ", ";

            currentNode = currentNode.getNext();
        }
        text += "]";

        System.out.println(text);
    }

    /**
     * 전체 요소 초기화
     */
    public void clear() {
        head = null;
        count = 0;
    }

    /**
     * 리스트 맨 뒤에 요소 추가하기
     */
    public void insertLast(T data) {
        insertAt(count, data);
    }

    /**
     * 맨 뒤에 요소 삭제하기
     */
    public Node<T> deleteLast() {
        return deleteAt(count - 1);
    }

    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }
}

File: theory/data_structure/LinkedList/python/linked_list.py
class Node:
    data = None
    next = None

    def __init__(self, data, next = None):
        self.data = data
        self.next = next


class LinkedList:
    head = None
    count = 0

    def insert_at(self, index, data):
        if index > self.count or index < 0:
            raise IndexError

        new_node = Node(data)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            current_node = self.head

            for i in range(index - 1):
                current_node = current_node.next
            new_node.next = current_node
            current_node.next = new_node
        self.count += 1

    def get_node_at(self, index):
        if index >= self.count or index < 0:
            raise IndexError

        current_node = self.head
        for i in range(index):
            current_node = current_node.next

        return current_node

    def print_all(self):
        current_node = self.head
        text = "["

        for i in range(self.count):
            text += str(current_node.data)

            if i is not self.count - 1:
                text += ", "

            current_node = current_node.next
        text += "]"
        print(text)

    def clear(self):
        self.head = None
        self.count = 0

    def insert_last(self, data):
        self.insert_at(self.count, data)

    def delete_at(self, index):
        if index >= self.count or index < 0:
            raise IndexError

        current_node = self.head

        if index == 0:
            delete_node = self.head
            self.head = self.head.next
            self.count -= 1

            return delete_node
        else:
            for i in range(index - 1):
                current_node = current_node.next

            delete_node = current_node.next
            current_node.next = current_node.next.next
            self.count -= 1

            return delete_node

    def delete_last(self):
        return self.delete_at(self.count - 1)


